### ISSUE TITLE
AI: Envoy fell down pit and cannot path an exit to map edge

### ISSUE THREAD

**Issue by com1clyf3** [CompletelyFairGames/dwarfcorp#719]
Trade envoy has fallen down pit dug around balloon port. Can't path its way out of the pit.

Save:
[Sea of Ghost_579_131708212430837856.zip](https://github.com/CompletelyFairGames/dwarfcorp/files/2003141/Sea.of.Ghost_579_131708212430837856.zip)

World:
[Sea of Ghost.zip](https://github.com/CompletelyFairGames/dwarfcorp/files/2003144/Sea.of.Ghost.zip)

5fee6751

**Comment by mklingen** [comment]
Yes, because it should be interpreted as the player trapping the envoy.

It does seem wrong that they can't climb out -- all intelligent creatures are supposed to be able to pathfind out of a pit. Were they coming to trade with you, or had they already traded?

**Comment by com1clyf3** [comment]
The envoy is not intentionally trapped. There are several surface level bridges across the pit the envoy could've used, but did not path across, instead choosing to go down into the pit.

They had already traded and were on their path to leave. Check the save if you want to observe.

**Comment by mklingen** [comment]
okay, will check the save. Not surprised they didn't find the bridges. The heuristic is tuned way way up to make planning as fast as possible.

**Comment by mklingen** [comment]
Ok, I just checked the save. Pathfinding is working fine, and trade envoys are making it in and out of the balloon port just fine. What we're looking at here are actually the *skeletons that a necromancer spawned*. It seems that once the original creatures that were in the trade envoy have left (they probably did this without you noticing), these skeletons are left behind. Going to try to reproduce with other trade envoys.

**Comment by mklingen** [comment]
Okay, so I tried a bunch of corner cases with trade envoys/ war parties and found a few interesting bugs that I've fixed. Some stuff that was happening:

- If any creature in a trade envoy had an invalid path for any reason (for example, if a block was mined along their path) they would give up and leave. I've now made them try 10 times before declaring failure. 
- If the path of a creature in a trade envoy was invalidated while it was leaving the world, the creature would freeze. There were a few things causing this. One was that timeouts in this case were set too high and weren't being counted correctly. Another is that the paths generated on failure were invalid. I've fixed that bug, and now the trade envoy will try to find an alternate route if their path gets invalidated. If there's no route, they will wander around randomly for about a minute before they are just disappeared from the world.
